File: sinain-mobile/android/app/src/main/java/com/isinain/pipeline/GatewayClient.kt
package com.isinain.pipeline

import kotlinx.coroutines.CancellableContinuation
import kotlinx.coroutines.suspendCancellableCoroutine
import okhttp3.*
import org.json.JSONArray
import org.json.JSONObject
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.Executors
import java.util.concurrent.ScheduledFuture
import java.util.concurrent.TimeUnit
import kotlin.coroutines.resume

/**
 * WebSocket gateway client for OpenClaw agent RPC.
 * Port of iOS GatewayClient.swift using OkHttp WebSocket.
 *
 * Protocol:
 *  1. Server sends connect.challenge event
 *  2. Client responds with connect request + auth token
 *  3. Client sends 'agent' RPC, server replies with accepted then final response
 *  4. Reconnect with exponential backoff
 *  5. Circuit breaker: 5 failures in 2min → open for 5min, progressive up to 30min
 */
class GatewayClient(
    private val wsUrl: String,
    private val token: String,
    private val sessionKey: String
) : GatewayConnecting {

    var onStatusChange: ((String) -> Unit)? = null
    var onResponse: ((String) -> Unit)? = null

    private val log = PipelineLogger("GatewayClient")
    private val httpClient = OkHttpClient.Builder()
        .readTimeout(0, TimeUnit.MILLISECONDS)  // no timeout for WebSocket reads
        .build()
    private val scheduler = Executors.newSingleThreadScheduledExecutor()

    private var webSocket: WebSocket? = null
    private var authenticated = false
    private var rpcId = 1
    private var closing = false
    private var pendingRunId: String? = null

    // Pending RPCs
    private data class PendingRpc(
        val continuation: CancellableContinuation<String?>,
        var timeoutFuture: ScheduledFuture<*>?,
        val expectFinal: Boolean
    )
    private val pending = ConcurrentHashMap<String, PendingRpc>()

    // Reconnect backoff
    private var reconnectDelay = 1.0
    private val maxReconnectDelay = 60.0
    private var reconnectFuture: ScheduledFuture<*>? = null

    // Circuit breaker
    private val recentFailures = mutableListOf<Long>()
    private var circuitOpenFlag = false
    private var circuitResetFuture: ScheduledFuture<*>? = null
    private val circuitThreshold = 5
    private val circuitWindowMs = 120_000L
    private var circuitResetDelay = 300_000L  // 5 min
    private val maxCircuitReset = 1_800_000L  // 30 min

    override val isConnected: Boolean
        get() = webSocket != null && authenticated

    override val isCircuitOpen: Boolean
        get() = circuitOpenFlag

    // ── Public API ────────────────────────────────────────────

    override fun start() {
        if (closing) return
        connect()
    }

    override suspend fun sendAgentRpc(message: String, idempotencyKey: String): String? {
        if (circuitOpenFlag) {
            log.warn("circuit breaker open, skipping RPC")
            return null
        }
        if (!isConnected) {
            log.warn("not connected, cannot send RPC")
            return null
        }

        val id = rpcId.toString()
        rpcId++

        return suspendCancellableCoroutine { continuation ->
            val timeoutFuture = scheduler.schedule({
                val rpc = pending.remove(id)
                if (rpc != null) {
                    onRpcFailure()
                    rpc.continuation.resume(null)
                }
            }, 60, TimeUnit.SECONDS)

            pending[id] = PendingRpc(continuation, timeoutFuture, true)

            continuation.invokeOnCancellation {
                pending.remove(id)?.timeoutFuture?.cancel(false)
            }

            pendingRunId = idempotencyKey

            val payload = JSONObject().apply {
                put("type", "req")
                put("method", "agent")
                put("id", id)
                put("params", JSONObject().apply {
                    put("message", message)
                    put("sessionKey", sessionKey)
                    put("idempotencyKey", idempotencyKey)
                    put("deliver", false)
                })
            }

            sendJson(payload)
            log.info("agent RPC sent (id=$id): ${message.take(100)}")
        }
    }

    override fun close() {
        closing = true
        reconnectFuture?.cancel(false)
        reconnectFuture = null
        circuitResetFuture?.cancel(false)
        circuitResetFuture = null
        webSocket?.close(1001, "going away")
        webSocket = null
        onDisconnect()
    }

    // ── Connection ────────────────────────────────────────────

    private fun connect() {
        if (closing || circuitOpenFlag) return
        onStatusChange?.invoke("connecting")
        log.info("connecting to $wsUrl")

        val request = Request.Builder().url(wsUrl).build()
        webSocket = httpClient.newWebSocket(request, object : WebSocketListener() {
            override fun onOpen(ws: WebSocket, response: Response) {
                log.info("ws connected (awaiting challenge)")
            }

            override fun onMessage(ws: WebSocket, text: String) {
                try {
                    val json = JSONObject(text)
                    handleMessage(json)
                } catch (e: Exception) {
                    log.error("failed to parse message: ${e.message}")
                }
            }

            override fun onFailure(ws: WebSocket, t: Throwable, response: Response?) {
                log.error("ws failure: ${t.message}")
                webSocket = null
                onDisconnect()
                if (!closing) scheduleReconnect()
            }

            override fun onClosing(ws: WebSocket, code: Int, reason: String) {
                log.info("ws closing: $code $reason")
                ws.close(code, reason)
            }

            override fun onClosed(ws: WebSocket, code: Int, reason: String) {
                log.info("ws closed: $code $reason")
                webSocket = null
                onDisconnect()
                if (!closing) scheduleReconnect()
            }
        })
    }

    // ── Message Handling ──────────────────────────────────────

    private fun handleMessage(msg: JSONObject) {
        val type = msg.optString("type", "")

        // 1. Handle connect.challenge
        if (type == "event" && msg.optString("event") == "connect.challenge") {
            log.info("received challenge, authenticating...")
            sendJson(JSONObject().apply {
                put("type", "req")
                put("id", "connect-1")
                put("method", "connect")
                put("params", JSONObject().apply {
                    put("minProtocol", 3)
                    put("maxProtocol", 3)
                    put("client", JSONObject().apply {
                        put("id", "gateway-client")
                        put("displayName", "ISinain Mobile (Android)")
                        put("version", "1.0.0")
                        put("platform", "android")
                        put("mode", "backend")
                    })
                    put("auth", JSONObject().apply {
                        put("token", token)
                    })
                })
            })
            return
        }

        // 2. Handle connect response (auth result)
        if (type == "res" && msg.optString("id") == "connect-1") {
            if (msg.optBoolean("ok", false)) {
                authenticated = true
                reconnectDelay = 1.0
                log.info("authenticated")
                onStatusChange?.invoke("connected")
            } else {
                val err = msg.opt("error") ?: "unknown auth error"
                log.error("auth failed: $err")
                reconnectDelay = maxOf(reconnectDelay, 30.0)
                webSocket?.close(1000, "auth failed")
                webSocket = null
            }
            return
        }

        // 3. Handle streaming events (filter by runId to prevent cross-talk)
        if (type == "event" && msg.optString("event") == "agent") {
            val payload = msg.optJSONObject("payload")
            if (payload != null) {
                // Only process events matching our pending RPC
                val eventRunId = payload.optString("runId", "")
                if (eventRunId != pendingRunId) return

                if (payload.optString("stream") == "assistant") {
                    val data = payload.optJSONObject("data")
                    val text = data?.optString("text", "") ?: ""
                    if (text.isNotEmpty()) {
                        onResponse?.invoke(text)
                    }
                }
            }
            return
        }

        // 4. Ignore other events
        if (type == "event") return

        // 5. Handle RPC responses
        val msgId: String? = when {
            msg.has("id") && msg.get("id") is Int -> msg.getInt("id").toString()
            msg.has("id") -> msg.optString("id")
            else -> null
        }

        if (type == "res" && msgId != null) {
            val pendingRpc = pending[msgId] ?: return
            val payload = msg.optJSONObject("payload") ?: JSONObject()

            // Skip intermediate "accepted" responses
            if (pendingRpc.expectFinal && payload.optString("status") == "accepted") {
                log.debug("rpc $msgId: accepted")
                return
            }

            log.debug("rpc $msgId: final")
            pendingRpc.timeoutFuture?.cancel(false)
            pending.remove(msgId)

            if (msg.optBoolean("ok", false)) {
                circuitResetDelay = 300_000L
                val result = payload.optJSONObject("result") ?: JSONObject()
                val payloads = result.optJSONArray("payloads") ?: JSONArray()
                val texts = mutableListOf<String>()
                for (i in 0 until payloads.length()) {
                    val text = payloads.optJSONObject(i)?.optString("text", "") ?: ""
                    if (text.isNotEmpty()) texts.add(text)
                }
                var responseText = texts.joinToString("\n")

                if (responseText.isEmpty()) {
                    val sentTexts = result.optJSONArray("messagingToolSentTexts")
                    if (sentTexts != null && sentTexts.length() > 0) {
                        val sentList = mutableListOf<String>()
                        for (i in 0 until sentTexts.length()) {
                            sentList.add(sentTexts.optString(i, ""))
                        }
                        responseText = sentList.joinToString("\n")
                    }
                }

                log.info("payloads: ${payloads.length()}, text: ${responseText.length} chars")
                if (responseText.isNotEmpty()) {
                    onResponse?.invoke(responseText)
                }
                pendingRpc.continuation.resume(responseText.ifEmpty { null })
            } else {
                val err = msg.opt("error") ?: "unknown RPC error"
                log.error("agent RPC error: $err")
                onRpcFailure()
                pendingRpc.continuation.resume(null)
            }
        }
    }

    // ── Helpers ───────────────────────────────────────────────

    private fun sendJson(payload: JSONObject) {
        webSocket?.send(payload.toString())
    }

    private fun onDisconnect() {
        authenticated = false
        for ((_, rpc) in pending) {
            rpc.timeoutFuture?.cancel(false)
            rpc.continuation.resume(null)
        }
        pending.clear()
        log.info("disconnected")
        onStatusChange?.invoke("disconnected")
    }

    private fun onRpcFailure() {
        val now = System.currentTimeMillis()
        recentFailures.add(now)
        val cutoff = now - circuitWindowMs
        recentFailures.removeAll { it < cutoff }

        if (recentFailures.size >= circuitThreshold && !circuitOpenFlag) {
            circuitOpenFlag = true
            log.warn("circuit breaker opened after ${recentFailures.size} failures, reset in ${circuitResetDelay / 1000}s")

            circuitResetFuture = scheduler.schedule({
                circuitOpenFlag = false
                recentFailures.clear()
                log.info("circuit breaker reset")
                if (!closing) connect()
            }, circuitResetDelay, TimeUnit.MILLISECONDS)

            circuitResetDelay = minOf(circuitResetDelay * 2, maxCircuitReset)
        }
    }

    private fun scheduleReconnect() {
        if (closing || circuitOpenFlag) return
        log.info("reconnecting in ${"%.1f".format(reconnectDelay)}s...")

        reconnectFuture = scheduler.schedule({
            reconnectFuture = null
            connect()
        }, (reconnectDelay * 1000).toLong(), TimeUnit.MILLISECONDS)

        reconnectDelay = minOf(reconnectDelay * 2, maxReconnectDelay)
    }
}

File: sinain-mobile/ios/ISinain/Pipeline/GatewayClient.swift
import Foundation

/// WebSocket gateway client for OpenClaw agent RPC.
///
/// Protocol:
///  1. Server sends connect.challenge event
///  2. Client responds with connect request + auth token
///  3. Client sends 'agent' RPC, server replies with accepted (intermediate) then final response
///  4. Reconnect with exponential backoff on disconnect
///  5. Circuit breaker: 5 failures in 2min -> open for 5min, progressive backoff up to 30min
final class GatewayClient: GatewayConnecting {
    typealias StatusCallback = (String) -> Void
    typealias ResponseCallback = (String) -> Void

    var onStatusChange: StatusCallback?
    var onResponse: ResponseCallback?

    private let wsUrl: String
    private let token: String
    private let sessionKey: String
    private let log = PipelineLogger(subsystem: "GatewayClient")

    private var urlSession: URLSession?
    private var wsTask: URLSessionWebSocketTask?
    private var authenticated = false
    private var rpcId = 1
    private var closing = false
    private var pendingRunId: String?

    // Pending RPCs
    private struct PendingRpc {
        let continuation: CheckedContinuation<String?, Never>
        var timer: DispatchWorkItem?
        let expectFinal: Bool
    }
    private var pending: [String: PendingRpc] = [:]

    // Reconnect backoff
    private var reconnectDelay: TimeInterval = 1.0
    private let maxReconnectDelay: TimeInterval = 60.0
    private var reconnectWorkItem: DispatchWorkItem?

    // Circuit breaker
    private var recentFailures: [Date] = []
    private var circuitOpenFlag = false
    private var circuitResetWorkItem: DispatchWorkItem?
    private let circuitThreshold = 5
    private let circuitWindowS: TimeInterval = 120
    private var circuitResetDelay: TimeInterval = 300
    private let maxCircuitReset: TimeInterval = 1800

    var isConnected: Bool {
        wsTask != nil && authenticated
    }

    var isCircuitOpen: Bool {
        circuitOpenFlag
    }

    init(wsUrl: String, token: String, sessionKey: String) {
        self.wsUrl = wsUrl
        self.token = token
        self.sessionKey = sessionKey
    }

    // MARK: - Public API

    func start() {
        guard !closing else { return }
        connect()
    }

    func sendAgentRpc(message: String, idempotencyKey: String) async -> String? {
        if circuitOpenFlag {
            log.warn("circuit breaker open, skipping RPC")
            return nil
        }
        guard isConnected else {
            log.warn("not connected, cannot send RPC")
            return nil
        }

        let id = String(rpcId)
        rpcId += 1

        return await withCheckedContinuation { continuation in
            let timeoutWork = DispatchWorkItem { [weak self] in
                guard let self = self, self.pending[id] != nil else { return }
                self.pending.removeValue(forKey: id)
                self.onRpcFailure()
                continuation.resume(returning: nil)
            }
            DispatchQueue.main.asyncAfter(deadline: .now() + 60, execute: timeoutWork)

            pending[id] = PendingRpc(
                continuation: continuation,
                timer: timeoutWork,
                expectFinal: true
            )

            self.pendingRunId = idempotencyKey

            let payload: [String: Any] = [
                "type": "req",
                "method": "agent",
                "id": id,
                "params": [
                    "message": message,
                    "sessionKey": sessionKey,
                    "idempotencyKey": idempotencyKey,
                    "deliver": false,
                ],
            ]

            sendJson(payload)
            log.info("agent RPC sent (id=\(id)): \(String(message.prefix(100)))")
        }
    }

    func close() {
        closing = true
        reconnectWorkItem?.cancel()
        reconnectWorkItem = nil
        circuitResetWorkItem?.cancel()
        circuitResetWorkItem = nil
        wsTask?.cancel(with: .goingAway, reason: nil)
        wsTask = nil
        urlSession?.invalidateAndCancel()
        urlSession = nil
        onDisconnect()
    }

    // MARK: - Connection

    private func connect() {
        guard !closing, !circuitOpenFlag else { return }
        onStatusChange?("connecting")
        log.info("connecting to \(wsUrl)")

        guard let url = URL(string: wsUrl) else {
            log.error("invalid URL: \(wsUrl)")
            scheduleReconnect()
            return
        }

        let session = URLSession(configuration: .default)
        urlSession = session
        let task = session.webSocketTask(with: url)
        wsTask = task
        task.resume()

        log.info("ws connected (awaiting challenge)")
        receiveMessage()
    }

    private func receiveMessage() {
        wsTask?.receive { [weak self] result in
            guard let self = self else { return }

            switch result {
            case .success(let message):
                switch message {
                case .string(let text):
                    if let data = text.data(using: .utf8),
                       let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] {
                        self.handleMessage(json)
                    }
                case .data(let data):
                    if let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] {
                        self.handleMessage(json)
                    }
                @unknown default:
                    break
                }
                self.receiveMessage()

            case .failure(let error):
                self.log.error("receive error: \(error)")
                self.wsTask = nil
                self.urlSession?.invalidateAndCancel()
                self.urlSession = nil
                self.onDisconnect()
                if !self.closing { self.scheduleReconnect() }
            }
        }
    }

    // MARK: - Message Handling

    private func handleMessage(_ msg: [String: Any]) {
        let type = msg["type"] as? String ?? ""

        // 1. Handle connect.challenge
        if type == "event", (msg["event"] as? String) == "connect.challenge" {
            log.info("received challenge, authenticating...")
            sendJson([
                "type": "req",
                "id": "connect-1",
                "method": "connect",
                "params": [
                    "minProtocol": 3,
                    "maxProtocol": 3,
                    "client": [
                        "id": "gateway-client",
                        "displayName": "ISinain Mobile (Native)",
                        "version": "1.0.0",
                        "platform": "ios",
                        "mode": "backend",
                    ] as [String: Any],
                    "auth": ["token": token],
                ] as [String: Any],
            ])
            return
        }

        // 2. Handle connect response (auth result)
        if type == "res", (msg["id"] as? String) == "connect-1" {
            if msg["ok"] as? Bool == true {
                authenticated = true
                reconnectDelay = 1.0
                log.info("authenticated")
                onStatusChange?("connected")
            } else {
                let err = msg["error"] ?? "unknown auth error"
                log.error("auth failed: \(err)")
                reconnectDelay = max(reconnectDelay, 30)
                wsTask?.cancel(with: .normalClosure, reason: nil)
                wsTask = nil
                urlSession?.invalidateAndCancel()
                urlSession = nil
            }
            return
        }

        // 3. Handle streaming events (filter by runId to prevent cross-talk)
        if type == "event", (msg["event"] as? String) == "agent" {
            if let payload = msg["payload"] as? [String: Any] {
                // Only process events matching our pending RPC
                let eventRunId = payload["runId"] as? String
                guard eventRunId == pendingRunId else { return }

                if (payload["stream"] as? String) == "assistant",
                   let data = payload["data"] as? [String: Any],
                   let text = data["text"] as? String {
                    onResponse?(text)
                }
            }
            return
        }

        // 4. Ignore other events
        if type == "event" { return }

        // 5. Handle RPC responses
        let msgId: String?
        if let intId = msg["id"] as? Int {
            msgId = String(intId)
        } else {
            msgId = msg["id"] as? String
        }

        if type == "res", let id = msgId, let pendingRpc = pending[id] {
            let payload = msg["payload"] as? [String: Any] ?? [:]

            // Skip intermediate "accepted" responses
            if pendingRpc.expectFinal, (payload["status"] as? String) == "accepted" {
                log.debug("rpc \(id): accepted")
                return
            }

            log.debug("rpc \(id): final")
            pendingRpc.timer?.cancel()
            pending.removeValue(forKey: id)

            if msg["ok"] as? Bool == true {
                circuitResetDelay = 300
                let result = payload["result"] as? [String: Any] ?? [:]
                let payloads = result["payloads"] as? [[String: Any]] ?? []
                let texts = payloads.compactMap { $0["text"] as? String }
                var responseText = texts.joined(separator: "\n")

                if responseText.isEmpty,
                   let sentTexts = result["messagingToolSentTexts"] as? [String], !sentTexts.isEmpty {
                    responseText = sentTexts.joined(separator: "\n")
                }

                log.info("payloads: \(payloads.count), text: \(responseText.count) chars")
                if !responseText.isEmpty {
                    onResponse?(responseText)
                }
                pendingRpc.continuation.resume(returning: responseText.isEmpty ? nil : responseText)
            } else {
                let err = msg["error"] ?? "unknown RPC error"
                log.error("agent RPC error: \(err)")
                onRpcFailure()
                pendingRpc.continuation.resume(returning: nil)
            }
            return
        }
    }

    // MARK: - Helpers

    private func sendJson(_ payload: [String: Any]) {
        guard let data = try? JSONSerialization.data(withJSONObject: payload),
              let text = String(data: data, encoding: .utf8) else { return }
        wsTask?.send(.string(text)) { error in
            if let error = error {
                self.log.error("send error: \(error)")
            }
        }
    }

    private func onDisconnect() {
        authenticated = false
        for (_, rpc) in pending {
            rpc.timer?.cancel()
            rpc.continuation.resume(returning: nil)
        }
        pending.removeAll()
        log.info("disconnected")
        onStatusChange?("disconnected")
    }

    private func onRpcFailure() {
        let now = Date()
        recentFailures.append(now)
        let cutoff = now.addingTimeInterval(-circuitWindowS)
        recentFailures = recentFailures.filter { $0 > cutoff }

        if recentFailures.count >= circuitThreshold && !circuitOpenFlag {
            circuitOpenFlag = true
            log.warn("circuit breaker opened after \(recentFailures.count) failures, reset in \(Int(circuitResetDelay))s")

            let resetWork = DispatchWorkItem { [weak self] in
                guard let self = self else { return }
                self.circuitOpenFlag = false
                self.recentFailures.removeAll()
                self.log.info("circuit breaker reset")
                if !self.closing { self.connect() }
            }
            circuitResetWorkItem = resetWork
            DispatchQueue.main.asyncAfter(deadline: .now() + circuitResetDelay, execute: resetWork)
            circuitResetDelay = min(circuitResetDelay * 2, maxCircuitReset)
        }
    }

    private func scheduleReconnect() {
        guard !closing, !circuitOpenFlag else { return }
        log.info("reconnecting in \(String(format: "%.1f", reconnectDelay))s...")

        let work = DispatchWorkItem { [weak self] in
            self?.reconnectWorkItem = nil
            self?.connect()
        }
        reconnectWorkItem = work
        DispatchQueue.main.asyncAfter(deadline: .now() + reconnectDelay, execute: work)
        reconnectDelay = min(reconnectDelay * 2, maxReconnectDelay)
    }
}
